src/System.Text.Encodings.Web/src/System/Text/Encodings/Web/TextEncoderExtensions.cs

```diff
@@ -0,0 +1,318 @@
+        {
+            if (value == null)
+            {
+                return value;
+            }
+
+            unsafe
+            {
+                fixed (char* valuePointer = value)
+                {
+                    int firstCharacterToEncode = encoder.FindFirstCharacterToEncode(valuePointer, value.Length);
+
+                    if (firstCharacterToEncode == -1)
+                    {
+                        return value;
+                    }
+
+                    int bufferSize = encoder.MaxOutputCharsPerInputChar * value.Length;
+                    char* wholebuffer = stackalloc char[bufferSize];
+                    char* buffer = wholebuffer;
+                    int totalWritten = 0;
+
+                    if (firstCharacterToEncode > 0)
+                    {
+                        int bytesToCopy = firstCharacterToEncode + firstCharacterToEncode;
+                        Buffer.MemoryCopy(valuePointer, buffer, bytesToCopy, bytesToCopy);
+                        totalWritten += firstCharacterToEncode;
+                        bufferSize -= firstCharacterToEncode;
+                        buffer += firstCharacterToEncode;
+                    }
+
+                    int valueIndex = firstCharacterToEncode;
+
+                    char firstChar = value[valueIndex];
+                    char secondChar = firstChar;
+                    bool wasSurrogatePair = false;
+                    int charsWritten;
+
+                    // this loop processes character pairs (in case they are surrogates).
+                    // there is an if block below to process single last character.
+                    for (int secondCharIndex = valueIndex + 1; secondCharIndex < value.Length; secondCharIndex++)
+                    {
+                        if (!wasSurrogatePair)
+                        {
+                            firstChar = secondChar;
+                        }
+                        else
+                        {
+                            firstChar = value[secondCharIndex - 1];
+                        }
+                        secondChar = value[secondCharIndex];
+
+                        if (!encoder.Encodes(firstChar))
+                        {
+                            wasSurrogatePair = false;
+                            *buffer = firstChar;
+                            buffer++;
+                            bufferSize--;
+                            totalWritten++;
+                        }
+                        else
+                        {
+                            int nextScalar = UnicodeHelpers.GetScalarValueFromUtf16(firstChar, secondChar, out wasSurrogatePair);
+                            if (!encoder.TryEncodeUnicodeScalar(nextScalar, buffer, bufferSize, out charsWritten))
+                            {
+                                throw new ArgumentException("encoder"); // the encoder MaxOutputCharsPerInputChar is not implemented correctly
```

Comment:
The `ArgumentException` ctor that takes one param takes it as a message, not as a parameter name: https://msdn.microsoft.com/en-us/library/wtxc6334(v=vs.110).aspx